analyzer/src/traversal/contracts.rs

```diff
@@ -26,12 +25,46 @@ use std::rc::Rc;
 /// Gather context information for contract definitions and check for type
 /// errors.
 pub fn contract_def(
+    module_scope: Shared<ModuleScope>,
+    _context: Shared<Context>,
+    stmt: &Spanned<fe::ModuleStmt>,
+) -> Result<(), SemanticError> {
+    if let fe::ModuleStmt::ContractDef { name, .. } = &stmt.node {
+        let contract_scope = ContractScope::new(name.node, Rc::clone(&module_scope));
+
+        contract_scope
+            .borrow()
+            .module_scope()
+            .borrow_mut()
+            .add_contract_def(
+                name.node,
+                Rc::clone(&contract_scope),
+                Contract {
+                    name: name.node.to_owned(),
+                    // Function defintions are added in a subsequent walk
```

Comment:
I'm thinking it would be cleaner to add the function definitions to the scope in this pass along with a new contract type definition.

If we've analyzed the signatures of the functions and added them to the scope, we should be able to create a contract type at the same time, eliminating the need for these partial contract definitions.

So:

`contract_def` pass
- adds the function definitions to the contract scope via a call `func_def`
- adds the contact type to the module scope

`contract_body` pass
- analyzes the body of each function in the contract via a call to `func_body`

Reply:
I had this in the beginning but there are two problems:

1. If we do that we lose the ability to use contract types in function signatures which on second thought I do think is valuable e.g. consider this:

```
def get strategy -> IStrategy:
    if some_condition:
        return FooStrategy(self.foo_address)
    else:
        return BarStrategy(self.bar_address)
```

If this is in contract `A` and `IStrategy` is defined below then we aren't aware of `IStrategy` at the point where we want to collect the function definitions from contract `A`.

2. Even if we ignore the former we will still wouldn't be able to obtain the final `ContractAttributes` after the first pass because `ContractAttributes` contains `external_contracts` so if contract `A` uses contract `B` (even if only in a function body) then the `ContractAttributes` of contract `A` need to also contain contract `B` which we haven't analysed at that point. In other words, even if we decide to collect the function definitions of contract `A` in the pass that creates the type, we will still need to recompute the `ContractAttributes` at the final hour when we have walked all the contracts so that they can contain all the `external_contracts` that are needed.

But as I write that down I noticed that this second issue isn't strongly related to the former. We could still do what you say and keep 2.) as is but I think we would lose out being able to handle the example code as explained in 1.)